File: 2022/day13/python/solution.py
#!/usr/bin/env python3
import os
import json
from functools import cmp_to_key

def compare(left, right):
    """
    Compare two values recursively.
    Returns: -1 if left < right (correct order)
              1 if left > right (wrong order)
              0 if equal (continue)
    """
    # Both integers
    if isinstance(left, int) and isinstance(right, int):
        if left < right:
            return -1
        elif left > right:
            return 1
        return 0

    # Both lists
    if isinstance(left, list) and isinstance(right, list):
        for i in range(min(len(left), len(right))):
            result = compare(left[i], right[i])
            if result != 0:
                return result
        # Check lengths
        if len(left) < len(right):
            return -1
        elif len(left) > len(right):
            return 1
        return 0

    # Mixed types - convert integer to list
    if isinstance(left, int):
        return compare([left], right)
    else:
        return compare(left, [right])

def part1(text):
    """Sum indices of pairs in correct order."""
    pairs = text.strip().split('\n\n')
    total = 0

    for i, pair in enumerate(pairs, 1):
        lines = pair.strip().split('\n')
        left = json.loads(lines[0])
        right = json.loads(lines[1])

        if compare(left, right) == -1:
            total += i

    return total

def part2(text):
    """Sort all packets with divider packets, find decoder key."""
    lines = [line for line in text.strip().split('\n') if line]
    packets = [json.loads(line) for line in lines]

    # Add divider packets
    divider1 = [[2]]
    divider2 = [[6]]
    packets.append(divider1)
    packets.append(divider2)

    # Sort using comparison function
    packets.sort(key=cmp_to_key(compare))

    # Find positions of dividers (1-indexed)
    pos1 = packets.index(divider1) + 1
    pos2 = packets.index(divider2) + 1

    return pos1 * pos2

def main():
    script_dir = os.path.dirname(os.path.abspath(__file__))
    input_file = os.path.join(script_dir, '..', 'input.txt')

    with open(input_file) as f:
        text = f.read()

    print('Part 1:', part1(text))
    print('Part 2:', part2(text))

if __name__ == '__main__':
    main()

File: 2022/day13/ruby/solution.rb
#!/usr/bin/env ruby
require 'json'

# Compare two values recursively.
# Returns: -1 if left < right (correct order)
#           1 if left > right (wrong order)
#           0 if equal (continue)
def compare(left, right)
  # Both integers
  if left.is_a?(Integer) && right.is_a?(Integer)
    return left <=> right
  end

  # Both arrays
  if left.is_a?(Array) && right.is_a?(Array)
    max_len = [left.length, right.length].min
    max_len.times do |i|
      result = compare(left[i], right[i])
      return result if result != 0
    end
    # Check lengths
    return left.length <=> right.length
  end

  # Mixed types - convert integer to array
  if left.is_a?(Integer)
    compare([left], right)
  else
    compare(left, [right])
  end
end

def part1(text)
  pairs = text.strip.split("\n\n")
  total = 0

  pairs.each_with_index do |pair, index|
    lines = pair.strip.split("\n")
    left = JSON.parse(lines[0])
    right = JSON.parse(lines[1])

    if compare(left, right) == -1
      total += index + 1  # 1-indexed
    end
  end

  total
end

def part2(text)
  lines = text.strip.split("\n").reject(&:empty?)
  packets = lines.map { |line| JSON.parse(line) }

  # Add divider packets
  divider1 = [[2]]
  divider2 = [[6]]
  packets << divider1
  packets << divider2

  # Sort using comparison function
  packets.sort! { |a, b| compare(a, b) }

  # Find positions of dividers (1-indexed)
  pos1 = packets.index(divider1) + 1
  pos2 = packets.index(divider2) + 1

  pos1 * pos2
end

def main
  script_dir = File.dirname(File.expand_path(__FILE__))
  input_file = File.join(script_dir, '..', 'input.txt')

  text = File.read(input_file)

  puts "Part 1: #{part1(text)}"
  puts "Part 2: #{part2(text)}"
end

main if __FILE__ == $PROGRAM_NAME
